packages/docusaurus/src/server/routes.ts

```diff
@@ -82,102 +118,36 @@ export default async function loadRoutes(
       );
     }
 
-    if (!routes) {
+    // Collect all page paths for injecting it later in the plugin lifecycle
+    // This is useful for plugins like sitemaps, redirects etc...
+    // If a route has subroutes, it is not necessarily a valid page path (more likely to be a wrapper)
+    if (!subroutes) {
       routesPaths.push(routePath);
     }
 
-    function genRouteChunkNames(
-      value: RouteModule | RouteModule[] | Module | null | undefined,
-      prefix?: string,
-      name?: string,
-    ) {
-      if (!value) {
-        return null;
-      }
-
-      if (Array.isArray(value)) {
-        return value.map((val, index) =>
-          genRouteChunkNames(val, `${index}`, name),
-        );
-      }
-
-      if (isModule(value)) {
-        const modulePath = getModulePath(value);
-        const chunkName = genChunkName(modulePath, prefix, name);
-        // We need to JSON.stringify so that if its on windows, backslashes are escaped.
-        const loader = `() => import(/* webpackChunkName: '${chunkName}' */ ${JSON.stringify(
-          modulePath,
-        )})`;
-
-        registry[chunkName] = {
-          loader,
-          modulePath,
-        };
-        return chunkName;
-      }
-
-      const newValue: ChunkNames = {};
-      Object.keys(value).forEach((key) => {
-        newValue[key] = genRouteChunkNames(value[key], key, name);
-      });
-      return newValue;
-    }
-
-    const alreadyExistingRouteChunkNames = routesChunkNames[routePath];
-    const chunkNames = {
-      ...genRouteChunkNames({component}, 'component', component),
-      ...genRouteChunkNames(modules, 'module', routePath),
+    // We hash the route to generate the key, because 2 routes can conflict with
+    // each others if they have the same path, ex: parent=/docs, child=/docs
+    // see https://github.com/facebook/docusaurus/issues/2917
+    const routeHash = simpleHash(JSON.stringify(routeConfig), 3);
+    const chunkNamesKey = `${routePath}-${routeHash}`;
+    routesChunkNames[chunkNamesKey] = {
+      ...genRouteChunkNames(registry, {component}, 'component', component),
+      ...genRouteChunkNames(registry, modules, 'module', routePath),
     };
-    // TODO is it safe to merge? that could lead to unwanted overrides
-    // See https://github.com/facebook/docusaurus/issues/2917
-    routesChunkNames[routePath] = {
-      ...alreadyExistingRouteChunkNames,
-      ...chunkNames,
-    };
-    if (alreadyExistingRouteChunkNames) {
-      console.warn(
-        chalk.red(
-          `It seems multiple routes have been created for routePath=[${routePath}], this can lead to unexpected behaviors.
-Components used for this route:
-- ${alreadyExistingRouteChunkNames.component}
-- ${chunkNames.component}
-${
-  routePath === '/'
-    ? "If you are using the docs-only/blog-only mode, don't forget to delete the homepage at ./src/pages/index.js"
-    : ''
-}
-`,
-        ),
```

Comment:
warning I put recently, can be removed now because conflits are unlikely to happen